backends/llamacpp/src/backend.rs

```diff
@@ -0,0 +1,423 @@
+            top_p: v.top_p,
+            frequency_penalty: v.frequency_penalty,
+            repetition_penalty: v.repetition_penalty,
+            seed: v.seed,
+        }
+    }
+}
+
+impl From<&ValidStoppingParameters> for GenerationParams {
+    fn from(v: &ValidStoppingParameters) -> Self {
+        Self {
+            max_new_tokens: v.max_new_tokens,
+            ignore_eos_token: v.ignore_eos_token,
+        }
+    }
+}
+
+#[cfg_attr(debug_assertions, derive(Debug))]
+pub(crate) struct GenerationContext {
+    pub(crate) input_tokens: Arc<Vec<u32>>,
+    pub(crate) generated_tokens: Vec<u32>,
+    pub(crate) generation_params: GenerationParams,
+    pub(crate) sampling_params: SamplingParams,
+}
+
+pub(crate) struct InferContext<'a> {
+    pub(crate) start: Instant,
+    pub(crate) stream: UnboundedSender<InferResult>,
+    pub(crate) tokenizer: &'a Tokenizer,
+    pub(crate) generation: GenerationContext,
+}
+
+#[derive(Debug, Error)]
+pub enum LlamaCppBackendError {
+    #[error("Provided GGUF model path {0} doesn't exist")]
+    ModelFileDoesntExist(String),
+
+    #[error("Failed to initialize model from GGUF file {0}: {1}")]
+    ModelInitializationFailed(PathBuf, String),
+}
+
+pub struct LlamaCppBackend {
+    scheduler_sender: UnboundedSender<(GenerationContext, UnboundedSender<InferResult>)>,
+    scheduler_handle: JoinHandle<()>,
+}
+
+impl LlamaCppBackend {
+    /// Attempt to create a new llama.cpp worker from the provided model path
+    ///
+    /// # Arguments
+    ///
+    /// * `path`: Path to the GGUF model file to load
+    /// * `num_threads`: Number of cores the model is allowed to spawn for its computations
+    ///
+    /// returns: Result<UniquePtr<LlamaCppWorkerFrontend>, LlamaCppBackendError>
+    ///
+    /// # Examples
+    ///
+    /// ```
+    ///
+    /// ```
+    fn allocate_worker(
+        path: &Path,
+        num_threads: u32,
+    ) -> Result<UniquePtr<LlamaCppWorkerFrontend>, LlamaCppBackendError> {
+        create_worker_frontend(&path.display().to_string(), num_threads).map_err(|ref err| {
+            LlamaCppBackendError::ModelInitializationFailed(path.to_path_buf(), err.to_string())
+        })
+    }
+
+    pub fn new<P: AsRef<Path>>(
+        model_path: P,
+        tokenizer: Arc<Tokenizer>,
+        num_cores_per_instance: u16,
+    ) -> Result<Self, LlamaCppBackendError> {
+        let path = model_path.as_ref();
+        if !path.exists() {
+            return Err(LlamaCppBackendError::ModelFileDoesntExist(
+                path.display().to_string(),
+            ));
+        }
+
+        // Allocate the multi-consumer queue to orchestrate all the workers
+        let (backlog_submitter, backlog_receiver) = mpmc_unbounded();
+
+        // Allocate all the workers
+        let cores_allocation = get_cores_allocation(num_cores_per_instance as usize);
+        cores_allocation.iter().for_each(|affinity| {
+            match Self::allocate_worker(path, num_cores_per_instance as u32) {
```

Comment:
You do not handle the case where `num_cores_per_instance` is not set as options -> it is then 0 and makes llama.cpp crash

Reply:
Fixed  in 298367cdfd2f33cc9699b2748f1d8edfbf3c66e1